docs/html/development/getting-started.rst

```diff
@@ -17,6 +17,16 @@ and ensuring reproducible environments for linting and building documentation.
 For developing pip, you need to install ``tox`` on your system. Often, you can
 just do ``python -m pip install tox`` to install and use it.
 
+Running pip From Source Tree
+----------------------------
+
+To run the pip executable from your source tree during development, run the
```

Comment:
Have you tried installing pip locally in editable mode? E.g.

```
$ pip install -e .
```

That seems like an easier way to run what's in your source directory (unless perhaps you mean a version of Python other than what is currently active in your environment).

Reply:
Using `python src/pip ...` would probably be the easiest way, once #5841 is merged.

Reply:
@benoit-pierre Would you rather use `python src/pip` already?

@cjerdonek That doesn't work consistently: e.g. when using a `conda` environment you will still get the `pip` version originally installed when the environment was created.

Reply:
It's probably an unintended consequence of the interaction between packages installed via `conda` and via `pip`/`setuptools` in the same conda environment. Maybe @kalefranz can shed some light.

<details>
<summary>Output of `pip freeze` and `conda list`</summary>

```
$ pip freeze
certifi==2018.10.15
filelock==3.0.9
pluggy==0.8.0
py==1.7.0
six==1.11.0
toml==0.10.0
tox==3.5.2
virtualenv==16.0.0
```

```
$ conda list
# packages in environment at /home/frg/local/miniconda/envs/pypa:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.6                h470a237_2    conda-forge
ca-certificates           2018.10.15           ha4d7672_0    conda-forge
certifi                   2018.10.15            py37_1000    conda-forge
filelock                  3.0.9                      py_0    conda-forge
libffi                    3.2.1                hfc679d8_5    conda-forge
libgcc-ng                 7.2.0                hdf63c60_3    conda-forge
libstdcxx-ng              7.2.0                hdf63c60_3    conda-forge
ncurses                   6.1                  hfc679d8_1    conda-forge
openssl                   1.0.2p               h470a237_1    conda-forge
pip                       19.0.dev0                 <pip>
pip                       18.1                  py37_1000    conda-forge
pluggy                    0.8.0                      py_0    conda-forge
py                        1.7.0                      py_0    conda-forge
python                    3.7.0                h5001a0f_4    conda-forge
readline                  7.0                  haf1bffa_1    conda-forge
setuptools                40.5.0                   py37_0    conda-forge
six                       1.11.0                py37_1001    conda-forge
sqlite                    3.25.2               hb1c47c0_0    conda-forge
tk                        8.6.8                ha92aebf_0    conda-forge
toml                      0.10.0                     py_0    conda-forge
tox                       3.5.2                      py_0    conda-forge
virtualenv                16.0.0                py37_1000    conda-forge
wheel                     0.32.2                   py37_0    conda-forge
xz                        5.2.4                h470a237_1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
```
</details

Reply:
(Btw, just to clarify, I'm not suggesting `pip install -e .` needs to be added. Just am curious to learn more.)

Reply:
Sure. I'm also curious to understand it myself :) So we're waiting for @benoit-pierre's review?

Reply:
The tox environments have pip installed as an editable: 

https://github.com/pypa/pip/blob/master/tox.ini#L25

I'd suggested this since it seemed to be something that others want to do often and the tox environments are usually set up the exact same way.

`python src/pip` seems perfectly reasonable as well (I just merged #5841), but I do prefer that we document a way here that uses an installed pip in a virtualenv; which the `.tox` directories are.

Reply:
For people who want to quickly run from source (e.g. to reproduce an issue, especially since they need access to the original environment), using `python src/pip` is really better. Having to run tox first means: you either need to suffer through the testsuite once (it's not fast), or know when to interrupt it or know to use `tox -e py --notest`, and again, all of that is of no use if you need to run pip with access to another environment.

Reply:
@msarahan See discussion above, do you have any thoughts?

Reply:
It's been a while, but I believe the last time I hacked on pip, I used either ``pip install -e .`` or ``python setup.py develop``.  That was version 9, IIRC, and I hit a very serious bit of confusion around the old, non-src layout, where I was getting a local pip import when pip (or maybe virtualenv?) really wanted its vendored wheel.

If you have specific repeatable examples of things not working with a conda-based env, I can look.